### DISCUSSION TITLE
Feathers js - how to login by google

### DISCUSSION THREAD

**Discussion by hhchow7** [feathersjs/feathers#2306] [Q&A]
I am following the [doc][1] to create google strategy.<br>

When I access http://localhost:3030/oauth/google through browser, below error occurs:

```
Error Code 400： redirect_uri_mismatch
The redirect URI in the request, https://localhost/oauth/google/callback, 
does not match the ones authorized for the OAuth client. 
To update the authorized redirect URIs, visit: https://console.developers.google.com/apis/credentials/oauthclient/12345678-xxxx.apps.googleusercontent.com?project=xxxxyyyy

```

[![enter image description here][2]][2]

authentication.js
```
const { AuthenticationService, AuthenticationBaseStrategy, JWTStrategy } = require('@feathersjs/authentication');
const { LocalStrategy } = require('@feathersjs/authentication-local');
const { expressOauth } = require('@feathersjs/authentication-oauth');

const axios = require('axios');
const { OAuthStrategy } = require('@feathersjs/authentication-oauth');

class GoogleStrategy extends OAuthStrategy {
  async getEntityData(profile) {
    const baseData = await super.getEntityData(profile);
    // this will grab the picture and email address of the Google profile
    return {
      ...baseData,
      email: profile.email
    };
  }
}

module.exports = app => {
  const authentication = new AuthenticationService(app);

  authentication.register('jwt', new JWTStrategy());
  authentication.register('local', new LocalStrategy());
  authentication.register('google', new GoogleStrategy());
 
  app.use('/authentication', authentication);
  app.configure(expressOauth());
};

```

config/local.json
```
{
    "authentication": {
      "entity": "user",
      "service": "users",
      "secret": "SA3c59SscyH6TscABCdeFG=",
      "authStrategies": [
        "jwt",
        "local",
        "google"
      ],
      "jwtOptions": {
        "header": {
          "typ": "access"
        },
        "audience": "https://yourdomain.com",
        "issuer": "feathers",
        "algorithm": "HS256",
        "expiresIn": "1d"
      },
      "local": {
        "usernameField": "email",
        "passwordField": "password"
      },
      "oauth": {
        "google": {
          "key": "xxx.apps.googleusercontent.com",
          "secret": "ASDFGgh"
        }
      }
    },
}
```

**Update 1**<br>
Fixed issue by adding `https://localhost/oauth/google/callback` in `Authorized redirect URIs`.<br>

Now the website redirect to `Select Account` Page. 

After I click my account, the website is redirected to `https://localhost/oauth/google/callback?code=4/abcd-xxx-xxxxx-xxx&scope=email%20profile%20https://www.googleapis.com/auth/userinfo.email%20https://www.googleapis.com/auth/userinfo.profile%20openid&authuser=0&hd=myDomain.com&prompt=consent`<br>

[![enter image description here][3]][3]

**Update 2<br>**

Updated local.json and add `redirect_uri` field
```
      "oauth": {
        "redirect": "/",
        "google": {
          "redirect_uri": "http://localhost:3030/auth/google/callback", // add here
          "key": "abcd.googleusercontent.com",
          "secret": "xxxx",
          "scope": [
            "email",
            "profile",
            "openid"
          ],
          "nonce": true
        }
      }
```
It can now redirect to `http://localhost:3030/auth/google/callback?code=xxx&scope=email+profile+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.email+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.profile+openid&authuser=0&hd=myDomain.com&prompt=consent#`

It shows `404 Page Not Find` instead of redirecting to `/`
[![enter image description here][4]][4]

**Update 3<br>**
Now the website redirects to `https://localhost/oauth/google/authenticate#`, I know it should be `https://localhost:3030/oauth/google/authenticate`, but I don't know how and where Feather can set it

**Update 4<br>**
Finally successfully redirecting to `www.google.com#error=Field%20password%20does%20not%20exist.%20(required)`,
is it kind of an error?
```
    "oauth": {
      "redirect": "www.google.com",
      "google": {
        "key": "<Google OAuth key>",
        "secret": "<Google OAuth secret>",
        "scope": [
          "email",
          "profile",
          "openid"
        ],
        "nonce": true,
        "redirect_uri": "http://localhost:3030/oauth/google/callback",
        "callback": "/oauth/google/authenticate"
      }
    }
```

  [1]: https://docs.feathersjs.com/cookbook/authentication/google.html#application-client-and-secret
  [2]: https://i.stack.imgur.com/2gqwK.png
  [3]: https://i.stack.imgur.com/O7jAb.png
  [4]: https://i.stack.imgur.com/bf9qQ.png

**Comment by rxb** [comment]
Most likely, password is a required field to create a user record, set up that way for the local strategy. 

What I've done is to populate a placeholder password for the newly OAuth'd (but not yet created in your app) user like this:

```
class GoogleStrategy extends OAuthStrategy {
  async getEntityData(profile) {
    const baseData = await super.getEntityData(profile);
    const newData = {
      ...baseData,
      name: profile.name,
      email: profile.email,
      password: Math.random().toString(36).substr(10) // or whatever type of temporary password you'd like
    };
    return newData;
  }
}
```
and then later

```
authentication.register('google', new GoogleStrategy());
```

The user will never use this specific password, it just exists as a placeholder known to no one. 
If you have password reset implemented in your app, the user could reset it if they later decide they want to use local auth.
